core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -36,9 +81,73 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
     IsolationLevel.READ_COMMITTED.toString.toLowerCase(Locale.ENGLISH)
   )
 
+  case object Drained
+  case class Drain[T](ack: ActorRef, msg: T)
+  case class Committed(offsets: Map[TopicPartition, OffsetAndMetadata])
+
   override protected def logic(shape: SourceShape[TransactionalMessage[K, V]]): GraphStageLogic with Control =
     new SingleSourceLogic[K, V, TransactionalMessage[K, V]](shape, txConsumerSettings, subscription)
     with TransactionalMessageBuilder[K, V] {
+      var inFlightRecords = InFlightRecords.empty
+
+      override def messageHandling = super.messageHandling.orElse(drainHandling).orElse {
+        case (_, Revoked(tps)) =>
+          inFlightRecords.revoke(tps.toSet)
+      }
+
+      override def shuttingDownReceive = super.shuttingDownReceive.orElse(drainHandling)
+
+      def drainHandling: PartialFunction[(ActorRef, Any), Unit] = {
+        case (sender, Committed(offsets)) =>
+          inFlightRecords.committed(offsets.mapValues(_.offset() - 1))
+          sender ! Done
+        case (_, Drain(ack, msg)) =>
+          if (inFlightRecords.empty()) {
+            log.debug("Source drained")
+            ack ! msg
```

Comment:
This should also send an answer back to the `sender` otherwise the `ask` on line 139 will not be completed.